[Admin] Add type hints to all files in the Admin dir

tambait opened the pull request:
<!-- THE PR TEMPLATE IS NOT AN OPTION. DO NOT DELETE IT, MAKE SURE YOU READ AND EDIT IT! -->
## Subject

This is a follow up of #6404 to add param & return type hints.

## Changelog

<!-- MANDATORY
    Fill the changelog part inside the code block.
    Follow this schema: http://keepachangelog.com/
    This will end up on https://github.com/sonata-project/SonataAdminBundle/releases,
    please keep it short and clear and to the point
-->

<!--
    If you are updating something that doesn't require
    a release, you can delete the whole "Changelog" section.
    (eg. update to docs, tests)
-->

<!-- REMOVE EMPTY SECTIONS -->
```markdown
### Added
- Added type hints to all files in the Admin dir.

VincentLanglet commented on src/Admin/AdminExtensionInterface.php:
```diff
@@ -155,13 +161,15 @@ public function configureActionButtons(
     /**
      * Returns a list of default filters.
      *
-     * @phpstan-param AdminInterface<object> $admin
```
should be kept

VincentLanglet commented on src/Admin/FieldDescriptionInterface.php:
```diff
@@ -18,36 +18,23 @@
-     *
-     * @return string the name
      */
     public function getName(): ?string;
```
Will be `:string` after the 3.X merge into master

VincentLanglet commented on src/Admin/FieldDescriptionInterface.php:
```diff
@@ -18,36 +18,23 @@
-    /**
-     * Set the name.
-     */
     public function setName(?string $name): void;
```
Should be `string $name`

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -2073,7 +2090,7 @@ public function getRouteBuilder(): ?RouteBuilderInterface
     }
 
-    public function toString(object $object): string
+    public function toString(?object $object): string
```
```suggestion
    public function toString(object $object): string
```

ToString should not allow null as param.

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -2580,7 +2597,7 @@ protected function buildForm(): void
     /**
      * Gets the subclass corresponding to the given name.
      *
-     * @phpstan-return class-string<T>
```
This seems weird

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -2676,16 +2693,15 @@ protected function configureDefaultFilterValues(array &$filterValues): void
     /**
      * Set the parent object, if any, to the provided object.
-     *
-     * @phpstan-param T $object
```
These removal seems weird

VincentLanglet commented on src/Admin/AdminInterface.php:
```diff
@@ -102,11 +104,12 @@ public function setTemplates(array $templates): void;
      */
     public function setTemplate(string $name, string $template): void;
 
+    public function setModelManager(ModelManagerInterface $modelManager): void;
```
Why is this needed ?

tambait replied:
Not sure tbh :), I saw `setModelManager` and 'setManagerType` additions in @jordisala1991  PR

VincentLanglet commented on src/Admin/AdminInterface.php:
```diff
@@ -102,11 +104,12 @@ public function setTemplates(array $templates): void;
-    /**
-     * @return string the manager type of the admin
-     */
+    public function setManagerType(string $managerType): void;
```
Why is this needed ?

VincentLanglet commented on src/Admin/AdminInterface.php:
```diff
@@ -231,7 +236,7 @@ public function getRouteBuilder(): ?RouteBuilderInterface;
      * @phpstan-param T $object
      */
-    public function toString(object $object): string;
+    public function toString(?object $object): string;
```
null shouldnt be allowed

VincentLanglet commented on src/Admin/BreadcrumbsBuilder.php:
```diff
@@ -70,7 +70,7 @@ public function getBreadcrumbs(AdminInterface $admin, string $action): array
      * Note: the method will be called by the top admin instance (parent => child)
      */
-    public function buildBreadcrumbs(
+    private function buildBreadcrumbs(
```
why this change ? This need a deprecation in 3.x.

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -2676,9 +2694,10 @@ protected function configureDefaultFilterValues(array &$filterValues): void
      *   $sortValues['_sort_by'] = 'foo'
      *   $sortValues['_sort_order'] = 'DESC'
      *
+     * @param array{_page?: int, _per_page?: int, _sort_by?: string, _sort_order?: string} $sortValues*
```
The detail is already present in phpstan-param, so we can remove this.

Or use array<string, string|int> which has a better integration by Phpstorm.

VincentLanglet commented on src/Admin/AdminExtensionInterface.php:
```diff
@@ -143,6 +145,12 @@ public function postRemove(AdminInterface $admin, object $object): void;
      *
+     * @param array<string, array<string, mixed>> $list
+     *
+     * @phpstan-param AdminInterface<object> $admin
```
The phpstan param is duplicated

VincentLanglet commented on src/Admin/AdminExtensionInterface.php:
```diff
@@ -155,13 +163,15 @@ public function configureActionButtons(
     /**
      * Returns a list of default sort values.
      *
+     * @param array{_page?: int, _per_page?: int, _sort_by?: string, _sort_order?: string} $sortValues
```
Similar. It would be better to use phpstan param for this

VincentLanglet commented on src/Admin/AdminHelper.php:
```diff
@@ -79,6 +75,8 @@ public function getChildFormView(FormView $formView, string $elementId): ?FormVi
      * @throws \RuntimeException
      * @throws \Exception
      *
+     * @return array{0: FieldDescriptionInterface|null, 1: FormInterface}
```
There is phpstan return for this

VincentLanglet commented on src/Admin/BaseFieldDescription.php:
```diff
@@ -119,7 +119,7 @@ abstract class BaseFieldDescription implements FieldDescriptionInterface
 
     /**
-     * @var array[] cached object field getters
+     * @var array<string, array{method: 'getter'|'call'|'var', getter?: string}>
```
```suggestion
     * @var string[][]
     * @phpstan-var array<string, array{method: 'getter'|'call'|'var', getter?: string}>
```

VincentLanglet commented on src/Admin/BaseFieldDescription.php:
```diff
@@ -438,6 +443,9 @@ private function callCachedGetter(object $object, string $fieldName, array $para
     }
 
+    /**
+     * @param 'call'|'getter'|'var' $method
```
```suggestion
     * @phpstan-param 'call'|'getter'|'var' $method
```

VincentLanglet commented on src/Admin/UrlGeneratorInterface.php:
```diff
@@ -68,7 +68,7 @@ public function generateUrl(string $name, array $parameters = [], int $reference
      * @param array<string, mixed> $parameters
      *
-     * @return array return url parts: 'route', 'routeParameters', 'routeAbsolute'
+     * @return array{route: string, routeParameters: array<string, string>, routeAbsolute: bool}
```
```suggestion
     * @return array<string, mixed>
     * @phpstan-return array{route: string, routeParameters: array<string, string>, routeAbsolute: bool}
```

tambait commented on src/Admin/AdminExtensionInterface.php:
```diff
@@ -155,7 +161,7 @@ public function configureActionButtons(
     /**
      * Returns a list of default filters.
      *
-     * @phpstan-param AdminInterface<object> $admin
```
Sorry, it seems I didn't push fixup for it

tambait commented on src/Admin/AdminHelper.php:
```diff
@@ -191,11 +187,10 @@ public function appendFormFieldElement(AdminInterface $admin, object $subject, s
      * @throws \Exception
      */
-    public function getElementAccessPath(string $elementId, $model): string
+    public function getElementAccessPath(string $elementId, object $model): string
```
I ported that from @jordisala1991  PR, so I'm not sure if it should be mixed or or there were some changes in the code. But, looking now at `isReadable` which takes `$model` as a first parameter, it expects it to be `@param object|array`

VincentLanglet replied:
Yeah, I know you got that from jordi PR. 

I just added a comment in order to check if it's right. (I was on phone).

The only usage is
```
$this->getElementAccessPath($elementId, $subject);
```
Where `$subject` is coming from `appendFormFieldElement`

But now I think that `getElementAccessPath` should be a private method since it's only used in AdminHelper.

VincentLanglet commented on tests/Admin/AdminTest.php:
```diff
@@ -745,8 +745,6 @@ public function testGetLabelTranslatorStrategy(): void
     {
         $admin = new PostAdmin('sonata.post.admin.post', 'NewsBundle\Entity\Post', 'Sonata\NewsBundle\Controller\PostAdminController');
 
-        $this->assertNull($admin->getLabelTranslatorStrategy());
```
I assume you remove this because it throws a TypeError.

IMHO, We should throw a LogicException in these getter when the value is null.
And we could then add a test with `$this->expectException(\LogicException::class)`.

But I'd like the @sonata-project/contributors point of view.
WDYT @jordisala1991 ?

jordisala1991 replied:
I would be a good idea, yes. I removed it to just make the test pass, but would be better to add a test instead

tambait replied:
I tried to do something, but not sure if I get the idea correctly. Also, LogicException messages could probably be more meaningful, I wasn't sure what to put so to be "generic" for now.

VincentLanglet replied:
You got the idea correctly. I juste made a minor comment on the exception message for consistency with others LogicException until we got something better.

VincentLanglet commented on tests/Admin/AdminHelperTest.php:
```diff
@@ -84,7 +84,11 @@ public function testGetElementAccessPath(): void
         $sub2Object->expects($this->atLeastOnce())->method('getMoreThings')->willReturn('Value');
 
-        $path = $this->helper->getElementAccessPath('uniquePartOfId_path_to_object_0_another_more_things', $object);
+        $path = $this->getMethodAsPublic('getElementAccessPath')->invoke(
```
I would say that since the method is now private, there is no need to test it (instead the test should be done by the method which call this private method, `appendFormFieldElement`).

But since there is currently no `appendFormFieldElement` tests, I think this is ok for now.

VincentLanglet commented:
```
1) Sonata\AdminBundle\Tests\Admin\Extension\LockExtensionTest::testPreUpdateIfAdminHasNoRequest
Method getModelManager may not return value of type NULL, its return declaration is ": Sonata\AdminBundle\Model\ModelManagerInterface"

2) Sonata\AdminBundle\Tests\Admin\Extension\LockExtensionTest::testPreUpdateIfObjectIsNotVersioned
Method getModelManager may not return value of type NULL, its return declaration is ": Sonata\AdminBundle\Model\ModelManagerInterface"

3) Sonata\AdminBundle\Tests\Admin\Extension\LockExtensionTest::testPreUpdateIfRequestDoesNotHaveLockVersion
Method getModelManager may not return value of type NULL, its return declaration is ": Sonata\AdminBundle\Model\ModelManagerInterface"

4) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testEditActionWithPreview
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

5) Sonata\AdminBundle\Tests\Controller\CRUDControllerTest::testCreateActionWithPreview
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

6) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtractEmpty
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

7) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtract
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

8) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtractCallsBreadcrumbs
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.

9) Sonata\AdminBundle\Tests\Translator\Extractor\AdminExtractorTest::testExtractSetsSubject
Class "Sonata\AdminBundle\Admin\FieldDescriptionCollection" is declared "final" and cannot be mocked.
```

The warning are easy to fix.

1) Change the mock of FieldDescriptionCollection by, `new FieldDescriptionCollection()`.

2) Mock the return value of `getModelManager` in `LockExtensionTest`, for this I would recommend to change 
```
    private function configureAdmin(
        string $uniqid = '',
        ?Request $request = null,
        ?ModelManagerInterface $modelManager = null
    ): void {
```
to
```
    private function configureAdmin(
        ModelManagerInterface $modelManager
        string $uniqid = '',
        ?Request $request = null,
    ): void {
```
in order to always set a ModelManager

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -1882,8 +1895,12 @@ public function setConfigurationPool(Pool $configurationPool): void
+    public function getConfigurationPool(): Pool
     {
+        if (null === $this->configurationPool) {
+            throw new \LogicException(sprintf('Pool has not been set.'));
```
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no pool.',
                static::class
            ));
```

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -1892,8 +1909,12 @@ public function setRouteGenerator(RouteGeneratorInterface $routeGenerator): void
+    public function getRouteGenerator(): RouteGeneratorInterface
     {
+        if (null === $this->routeGenerator) {
+            throw new \LogicException(sprintf('Router has not been set.'));
```
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no route generator.',
                static::class
            ));
```

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -1911,12 +1932,16 @@ public function getBaseCodeRoute(): string
+    public function getModelManager(): ModelManagerInterface
     {
+        if (null === $this->modelManager) {
+            throw new \LogicException(sprintf('Manager has not been set.'));
```
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no model manager.',
                static::class
            ));
```

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -2087,8 +2120,12 @@ public function setLabelTranslatorStrategy(LabelTranslatorStrategyInterface $lab
+    public function getLabelTranslatorStrategy(): LabelTranslatorStrategyInterface
     {
+        if (null === $this->labelTranslatorStrategy) {
+            throw new \LogicException(sprintf('LabelTranslatorStrategyInterface has not been set.'));
```
```suggestion
            throw new \LogicException(sprintf(
                'Admin "%s" has no label translator strategy.',
                static::class
            ));
```

VincentLanglet reviewed (commented):
You got the idea.

We already have message like
```
throw new \LogicException(sprintf(
                'Admin "%s" has no parent field description.',
                static::class
            ));
```

So I would go with something similar at beginning, until someone eventually improve the message to explain how to register/set all these services.

tambait commented:
> 1. Change the mock of FieldDescriptionCollection by, `new FieldDescriptionCollection()`.

Thanks. I'm not sure where to make this change, though maybe I was searching for/looking at wrong thing. When I searched through test files all `FieldDescriptionCollection` references were instantiated with `new FieldDescriptionCollection()` nor I found something like `CreateMock` etc

> 2. Mock the return value of `getModelManager` in `LockExtensionTest`, for this I would recommend to change

I've changed `configureAdmin` method but this caused errors in three other methods from the test class which didn't pass any parameters to original `configureAdmin` method or they passed just two parameters. I updated all of these to have `$this->configureAdmin('', null, $this->modelManager)` which removed the error, but not sure if this could have a unwanted impact on the test.

I also don't have idea how to solve phpStan errors :)

VincentLanglet commented on tests/Admin/Extension/LockExtensionTest.php:
```diff
@@ -235,7 +235,7 @@ private function configureFormMapper(): FormMapper
         string $uniqid = '',
         ?Request $request = null,
-        ?ModelManagerInterface $modelManager = null
+        ModelManagerInterface $modelManager
```
Can you change the order of the param to 
`$modelManager, $uniqId, $request`.

This allow `$this->configureAdmin($this->modelManager)` instead of `$this->configureAdmin('', null, $this->modelManager)`

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -869,6 +868,9 @@ public function getBaseRouteName(): string
         return $this->cachedBaseRouteName;
     }
 
+    /**
+     * @phpstan-return class-string
+     */
```
Remove this

VincentLanglet reviewed (commented):
Edit: 

For one of your issue with Phpstan, there is some conflict between 
```
   /**
     * Returns list of supported sub classes.
     *
     * @return array<string, string>
     *
     * @phpstan-return array<string, class-string>
     */
    public function getSubClasses(): array;

    /**
     * Sets the list of supported sub classes.
     *
     * @param array<string, string> $subClasses
     *
     * @phpstan-param array<string, class-string> $subClasses
     */
    public function setSubClasses(array $subClasses): void;
```
And
```
/**
     * Returns subjectClass/class/subclass name managed
     * - subclass name if subclass parameter is defined
     * - subject class name if subject is defined
     * - class name if not.
     *
     * @phpstan-return class-string<T>
     */
    public function getClass(): string;
```

I think `<T>` should be removed for `getClass()`.

And then, you have to update
```
/**
     * Gets the subclass corresponding to the given name.
     *
     * @phpstan-return class-string<T>
     */
    protected function getSubClass(string $name): string
```
to remove the `<T>` too.

VincentLanglet commented:
I would say you have to do
```
/**
     * The subclasses supported by the admin class.
     *
     * @var string[]
     *
     * @phpstan-var array<string, class-string<T>>
     */
    private $subClasses = [];
```
In AbstractAdmin

Then
```
/**
     * Returns list of supported sub classes.
     *
     * @return array<string, string>
     *
     * @phpstan-return array<string, class-string<T>>
     */
    public function getSubClasses(): array;

    /**
     * Sets the list of supported sub classes.
     *
     * @param array<string, string> $subClasses
     *
     * @phpstan-param array<string, class-string<T>> $subClasses
     */
    public function setSubClasses(array $subClasses): void;

    /**
     * Returns true if the admin has the sub classes.
     *
     * @phpstan-param class-string<T> $name
     */
    public function hasSubClass(string $name): bool;
```
In the AdminInterface

VincentLanglet commented on src/Admin/AbstractAdmin.php:
```diff
@@ -2676,9 +2725,10 @@ protected function configureDefaultFilterValues(array &$filterValues): void
      *   $sortValues['_sort_by'] = 'foo'
      *   $sortValues['_sort_order'] = 'DESC'
      *
+     * @param array<string, string|int>
```
```suggestion
     * @param array<string, string|int> $sortValues
     *
```

VincentLanglet commented:
For the phpstan error inside attacheInlineValidator, I would recommend to write
```
// @phpstan-ignore-next-line
$admin->validate($errorElement, $object);
```
It should be removed: https://github.com/sonata-project/SonataAdminBundle/issues/6233

VincentLanglet commented:
Do you encounter some issues or is it just a question of time ?
I don't want to hurry at all.

I just wanted to tell you that if you encounter issues and need help, don't hesitate to ping me. ;)
I'm working on multiple PR and don't always think to take a look at yours.

tambait commented:
> Do you encounter some issues or is it just a question of time ?
> I don't want to hurry at all.
> 
> I just wanted to tell you that if you encounter issues and need help, don't hesitate to ping me. ;)
> I'm working on multiple PR and don't always think to take a look at yours.

Thanks. Little bit of both :) I understand that you're working on multiple PR's so I was looking to solve issues that remained by myself. If I didn't oversee something think that "major" problems are with FieldDescriptionCollection in [tests](https://github.com/sonata-project/SonataAdminBundle/pull/6539#issuecomment-717432398) as well as phpStan conflict that you commented [here](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831).

VincentLanglet commented:
> Thanks. Little bit of both :) I understand that you're working on multiple PR's so I was looking to solve issues that remained by myself. If I didn't oversee something think that "major" problems are with FieldDescriptionCollection in [tests](https://github.com/sonata-project/SonataAdminBundle/pull/6539#issuecomment-717432398) as well as phpStan conflict that you commented [here](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831).

Great.

I think I gave everything needed for the phpstan issue.

For the fieldDescriptionCollection, it's weird indeed. I saw no `createMock()` calls.
But for instance, `testCreateActionWithPreview` will execute this code in the createAction
```
elseif ($this->isPreviewRequested()) {
                // pick the preview template if the form was valid and preview was requested
                $templateKey = 'preview';
                $this->admin->getShow();
            }
```

Since the `admin` is a mock in the tests (see the setup), Phpunit automatically create a mock for the return value.
Before the type was `?FieldDescriptionCollection` so it was returning null.
Now you restrict to `FieldDescriptionCollection` so it will mock FieldDescriptionCollection.

You should then, in the setup of CRUDControllerTest, add 
```
$this->admin->method('getShow')->willReturn(new FieldDescriptionCollection())
```

Same for others issues (maybe ?)

tambait commented:
> I think I gave everything needed for the phpstan issue.

Yes you did. All phpStan fixes are implemented except, I think, the one you [asked](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831) @core23 about

It's the error on line 39 in `FieldDescriptionInterface.php` which states `PHPDoc tag @return with type string|null is not subtype of native type string.`

>Since the admin is a mock in the tests (see the setup), Phpunit automatically create a mock for the return value.
Before the type was ?FieldDescriptionCollection so it was returning null.
Now you restrict to FieldDescriptionCollection so it will mock FieldDescriptionCollection.`

> You should then, in the setup of CRUDControllerTest, add
`$this->admin->method('getShow')->willReturn(new FieldDescriptionCollection())`


Great. I was trying to do something similar with `willReturn` but in a wrong way. For the change you suggested, if I put the code inside setUp method it triggers two failures `Failed asserting that two variables reference the same object` in

> 1) CRUDControllerTest::testHistoryViewRevisionAction :2969
> 2) CRUDControllerTest::testHistoryCompareRevisionsActionAction :3184

however if I put the the line inside, for example, `testEditActionWithPreview` method it goes without failures. I guess I could do the same for all methods which produce warning in the test, but I'm not sure if placing the line in setUp method and fixing new failures could produce same results without additional side effects.

VincentLanglet commented:
> > I think I gave everything needed for the phpstan issue.
> 
> Yes you did. All phpStan fixes are implemented except, I think, the one you [asked](https://github.com/sonata-project/SonataAdminBundle/pull/6539#pullrequestreview-518036831) @core23 about
> 
> It's the error on line 39 in `FieldDescriptionInterface.php` which states `PHPDoc tag @return with type string|null is not subtype of native type string.`

You can follow: https://github.com/sonata-project/SonataAdminBundle/pull/6539#issuecomment-717472455

> > Since the admin is a mock in the tests (see the setup), Phpunit automatically create a mock for the return value.
> > Before the type was ?FieldDescriptionCollection so it was returning null.
> > Now you restrict to FieldDescriptionCollection so it will mock FieldDescriptionCollection.`
> 
> > You should then, in the setup of CRUDControllerTest, add
> > `$this->admin->method('getShow')->willReturn(new FieldDescriptionCollection())`
> 
> Great. I was trying to do something similar with `willReturn` but in a wrong way. For the change you suggested, if I put the code inside setUp method it triggers two failures `Failed asserting that two variables reference the same object` in
> 
> > 1. CRUDControllerTest::testHistoryViewRevisionAction :2969
> > 2. CRUDControllerTest::testHistoryCompareRevisionsActionAction :3184
> 
> however if I put the the line inside, for example, `testEditActionWithPreview` method it goes without failures. I guess I could do the same for all methods which produce warning in the test, but I'm not sure if placing the line in `setUp` method and fixing new failures could produce same results without additional side effects.

It's because the method is already mocked.
I would say then that it's easier to put the line inside `testEditActionWithPreview` instead of `setUp`.

tambait commented:
@VincentLanglet 

PHPUnit Tests are hopefully good now, but I'm still puzzled with 

>  Line 39   FieldDescriptionInterface.php                                                   
>   PHPDoc tag @return with type string|null is not subtype of native type string.  

If I run phpstan locally on my branch it goes without error. If locally run phpstan on master branch I get same error check as above. Master branch has

```
    /**
     * NEXT_MAJOR: Restrict return type to string.
     *
     * Returns the name, the name can be used as a form label or table header.
     *
     * @return string|null the name
     */
    public function getName(): string;
```
while in my branch I have

```
    /**
     * Returns the name, the name can be used as a form label or table header.
     *
     * @return string the name
     */
    public function getName(): string;
```
When I look at annotation about the error, I have feeling like it's displaying code from my branch but complaining about the code from the master branch :smiley: 

There are few errors reported by psalm

```
Error: Class or interface Symfony\Bundle\MakerBundle\Maker\AbstractMaker does not exist
Error: Class or interface Symfony\Component\Yaml\Yaml does not exist
Error: Sonata\AdminBundle\Maker\AdminMaker depends 
on class or interface symfony\bundle\makerbundle\maker\abstractmaker that does not exist

```
and there was also
```
$this->securityInformation with declared type 'array<string, array<array-key, string>>' 
cannot be assigned type 'array<string, array<array-key, string>|string>'
```
where I changed declaration for `$securityInformation` from 
`@var array<string, string[]> 'role' => ['permission', 'permission']`

to 

`@var array<string, string[]|string> 'role' => ['permission', 'permission']`

VincentLanglet commented:
I merged 3.x into Master

Wait for https://github.com/sonata-project/SonataAdminBundle/pull/6557, maybe it will fix your issues.

tambait commented:
It could be that something was messed during resolving conflicts with earlier version of master. I now modified phpdoc block by removing `@return string the name`, resolved the conflict with the  (updated) master and phpstan error disappeared
```
    /**
     * Returns the name, the name can be used as a form label or table header.
     *
     * @return string the name
     */
     public function getName(): string
```

VincentLanglet commented:
Please review @sonata-project/contributors

SonataCI commented:
Could you please rebase your PR and fix merge conflicts?

VincentLanglet commented:
@tambait Do you have some time to rebase ? Then I'll merge it.

tambait commented:
> @tambait Do you have some time to rebase ? Then I'll merge it.

Yes, thank for notifying me (I missed notification about conflicts). 
After this one is merged, is there some "to do" list or something similar I can look at for future work/contributions? At least for me personally, list you've made for param/return type hints was very useful to get involved with the project

VincentLanglet commented on src/Admin/Pool.php:
```diff
@@ -63,32 +57,28 @@ class Pool
-     * @var MutableTemplateRegistryInterface
+     * @param array<string, mixed> $options
      */
-    private $templateRegistry;
```
Seems like this should not be removed

tambait replied:
Yes, thanks. Not sure if I messed up something. Should I also replace MutableTemplateRegistryInterface with TemplateRegistryInterfac?  Is that comment is for 4.0 version (not sure I saw it before)

VincentLanglet commented:
You messed up the phpdoc with your fix up

tambait commented:
> You messed up the phpdoc with your fix up

Oh. I just run php-cs-fixer. How can I fix that?

VincentLanglet commented:
It should be
```
/**
  * NEXT_MAJOR: change to TemplateRegistryInterface.
  *
  * @var MutableTemplateRegistryInterface
  */
private $templateRegistry;
```
because
```
     * @param array<string, mixed> $options
```
is related to the constructor.

tambait commented:
> It should be
> 
> ```
> /**
>   * NEXT_MAJOR: change to TemplateRegistryInterface.
>   *
>   * @var MutableTemplateRegistryInterface
>   */
> private $templateRegistry;
> ```
> 
> because
> 
> ```
>      * @param array<string, mixed> $options
> ```
> 
> is related to the constructor.

Thanks. Just realized what was problem

VincentLanglet reviewed (commented):
I think this https://github.com/sonata-project/SonataAdminBundle/pull/6539/files#diff-0850515af24b81ce716f53b8eb0b71fae91dcbdcc88b28d95a83886504e1c8f4R329

Should be Pool|null

VincentLanglet commented:
Anyway I'll work on the `null` issue in https://github.com/sonata-project/SonataAdminBundle/pull/6614

VincentLanglet commented:
Thanks @tambait.

If you want to continue helping us, I would recommend the milestone 4.0
https://github.com/sonata-project/SonataAdminBundle/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.0

Don't hesitate to ask anything before starting an issue. I already start working on some.

tambait commented:
> Thanks @tambait.
> 
> If you want to continue helping us, I would recommend the milestone 4.0
> https://github.com/sonata-project/SonataAdminBundle/issues?q=is%3Aopen+is%3Aissue+milestone%3A4.0
> 
> Don't hesitate to ask anything before starting an issue. I already start working on some.

Great thanks, I'll take a look.
